catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/app/FormulaEditorEditTextTest.java

```diff
@@ -0,0 +1,407 @@
+				.check(matches(withText("99" + decimalMark + "9 ")));
+		onView(allOf(withId(R.id.brick_change_size_by_edit_text), withEffectiveVisibility(Visibility.VISIBLE)))
+				.perform(click());
+		pressBack();
+
+		onView(allOf(withId(R.id.brick_change_size_by_edit_text), withEffectiveVisibility(Visibility.VISIBLE)))
+				.check(matches(withText("99" + decimalMark + "9 ")));
+
+		onView(allOf(withId(R.id.brick_change_size_by_edit_text), withEffectiveVisibility(Visibility.VISIBLE)))
+				.perform(click());
+		onView(withId(R.id.formula_editor_keyboard_delete))
+				.perform(click());
+		onView(withId(R.id.formula_editor_keyboard_delete))
+				.perform(click());
+		pressBack();
+		onView(withText(getCurrentActivity().getString(R.string.formula_editor_discard_changes_dialog_title)))
+				.check(matches(isDisplayed()));
+		onView(withText(no))
+				.perform(click());
+
+		checkToastMessage(R.string.formula_editor_changes_discarded);
+		onView(allOf(withId(R.id.brick_change_size_by_edit_text), withEffectiveVisibility(Visibility.VISIBLE)))
+				.check(matches(withText("99" + decimalMark + "9 ")));
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testFormulaIsNotValidToast() {
+		onView(withId(R.id.formula_editor_keyboard_delete))
+				.perform(click());
+		onView(withId(R.id.formula_editor_keyboard_ok))
+				.perform(click());
+		checkToastMessage(R.string.formula_editor_parse_fail);
+
+		onView(withId(R.id.formula_editor_keyboard_delete))
+				.perform(click());
+		onFormulaEditor()
+				.performEnterFormula("+");
+		onView(withId(R.id.formula_editor_keyboard_ok))
+				.perform(click());
+		checkToastMessage(R.string.formula_editor_parse_fail);
+
+		onView(withId(R.id.formula_editor_keyboard_delete))
+				.perform(click());
+		onFormulaEditor()
+				.performEnterFormula("1+1+");
+		pressBack();
+		onView(withText(getCurrentActivity().getString(R.string.formula_editor_discard_changes_dialog_title)))
+				.check(matches(isDisplayed()));
+		onView(withText(yes))
+				.perform(click());
+		checkToastMessage(R.string.formula_editor_parse_fail);
+	}
+
+	@Category({Cat.CatrobatLanguage.class, Level.Smoke.class})
+	@Test
+	public void testComputeDialog() {
+		onFormulaEditor()
+				.performEnterFormula("-2");
+		onView(withId(R.id.formula_editor_keyboard_compute))
+				.perform(click());
+		onView(withId(R.id.formula_editor_compute_dialog_textview))
+				.check(matches(withText("-2.0")));
+		pressBack();
+
+		onFormulaEditor()
+				.performEnterFormula("-6.111-");
+		onView(withId(R.id.formula_editor_keyboard_compute))
+				.perform(click());
+		onView(withId(R.id.formula_editor_keyboard_delete))
+				.perform(click());
+		onView(withId(R.id.formula_editor_keyboard_compute))
+				.perform(click());
+		onView(withId(R.id.formula_editor_compute_dialog_textview))
+				.check(matches(withText("-8.111")));
+	}
+
+	private int getAbsoluteCursorPosition() {
```

Comment:
it would be even nicer if youd put this in a custom matcher.
and then add this to the formula editor wrapper

Reply:
do you mean I should add a function like this:
```
public FormulaEditorWrapper checkHasCursorPosition(int expected)
{
    FormulaEditorEditText editText = (FormulaEditorEditText) getCurrentActivity().findViewById(R.id.formula_editor_edit_field);
    Object position = Reflection.getPrivateField(editText, "absoluteCursorPosition");
    assertEquals(((Number) position).intValue(),expected);
    return new FormulaEditorWrapper();
} 
```
to the FormulaEditorWrapper? This seems somehow wrong to me since I directly access private fields of the FormulaEditorEditText and this isn't really Espresso stuff and I have no idea how to do this in Espresso or if this is even possible to do